pkg/trace/writer/newwriter/trace.go

```diff
@@ -0,0 +1,263 @@
+package writer
+
+import (
+	"bytes"
+	"compress/gzip"
+	"math"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/trace/info"
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics/timing"
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+	"github.com/DataDog/datadog-agent/pkg/trace/traceutil"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"github.com/gogo/protobuf/proto"
+)
+
+// pathTraces is the target host API path for delivering traces.
+const pathTraces = "/api/v0.2/traces"
+
+// payloadFlushThreshold specifies the maximum accumulated payload size that is allowed before
+// a flush is triggered; replaced in tests.
+var payloadFlushThreshold = 3200000 // 3.2MB is the maximum allowed by the Datadog API
+
+// SampledSpans represents the result of a trace sampling operation.
+type SampledSpans struct {
+	// Trace will contain a trace if it was sampled or be empty if it wasn't.
+	Trace pb.Trace
+	// Events contains all APM events extracted from a trace. If no events were extracted, it will be empty.
+	Events []*pb.Span
+}
+
+// Empty returns true if this TracePackage has no data.
+func (ss *SampledSpans) Empty() bool {
+	return len(ss.Trace) == 0 && len(ss.Events) == 0
+}
+
+// size returns the estimated size of the package.
+func (ss *SampledSpans) size() int {
+	// we use msgpack's Msgsize() heuristic because it is a good indication
+	// of the weight of a span and the msgpack size is relatively close to
+	// the protobuf size, which is expensive to compute.
+	return ss.Trace.Msgsize() + pb.Trace(ss.Events).Msgsize()
+}
+
+type TraceWriter struct {
+	in       <-chan *SampledSpans
+	hostname string
+	env      string
+	senders  []*sender
+	stop     chan struct{}
+
+	stats *info.TraceWriterInfo
+
+	traces       []*pb.APITrace // traces buffered
```

Comment:
is this thread-safe between add and flush?

Reply:
Yes, the `select` statement synchronizes access. The "bigger" changes are in `sender.go`, this file didn't change that much from what we had.